upf/upf_pfcp_api.c

```diff
@@ -1669,7 +1702,43 @@ handle_update_far (upf_session_t * sx, pfcp_update_far_t * update_far,
 	    ip_udp_gtpu_rewrite (&update->forward, fib_index, is_ip4);
 	  }
 	//TODO: transport_level_marking
-	//TODO: forwarding_policy
+	//forwarding_policy  >> oln: Implementation
+	if (ISSET_BIT (far->update_forwarding_parameters.grp.fields,
+		       UPDATE_FORWARDING_PARAMETERS_FORWARDING_POLICY))
+	  {
+	    policy_id =
+	      far->update_forwarding_parameters.forwarding_policy.identifier;
+	    hash_ptr = hash_get_mem (gtm->forwarding_policy_by_id, policy_id);
+	    if (hash_ptr)
+	      {
+		if (update->forward.fp_pool_index != ~0)	// old policy is preconfigured
+		  {
+		    fp_entry = pool_elt_at_index (gtm->upf_forwarding_policies, update->forward.fp_pool_index);	// old policy
+		    fp_entry->ref_cnt--;	// decrement FAR reference counter in old policy table
+		    fp_entry = pool_elt_at_index (gtm->upf_forwarding_policies, hash_ptr[0]);	// new policy
+		    update->forward.forwarding_policy.identifier =
+		      far->update_forwarding_parameters.
+		      forwarding_policy.identifier;
+		    update->forward.fp_pool_index = hash_ptr[0];	// update with new pool index
+		    fp_entry->ref_cnt++;	// increment FAR reference counter in new policy table
+		  }
+		else		// old policy was not preconfigured somehow but we still update to new policy
+		  {
+		    fp_entry = pool_elt_at_index (gtm->upf_forwarding_policies, hash_ptr[0]);	// new policy
+		    update->forward.flags |= FAR_F_FORWARDING_POLICY;
+		    update->forward.forwarding_policy.identifier =
+		      far->update_forwarding_parameters.
+		      forwarding_policy.identifier;
+		    update->forward.fp_pool_index = hash_ptr[0];	// set pool index
+		    fp_entry->ref_cnt++;	// increment FAR reference counter in new policy table
+		  }
+	      }
+	    else
+	      upf_debug
+		("###### Forwarding policy id %v is not preconfigured at UPF ######",
+		 far->update_forwarding_parameters.
+		 forwarding_policy.identifier);
+	  }
```

Comment:
here also, reject request and return proper error cause

Reply:
Done